internal/webhook/external/server.go

```diff
@@ -71,3 +75,108 @@ func (s *server) Serve(ctx context.Context, l net.Listener) error {
 		return err
 	}
 }
+
+func (s *server) refreshWarehouseHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := (r.Context())
+	logger := logging.LoggerFromContext(ctx)
+	logger.Debug("refresh warehouse handler called", "path", r.URL.Path)
+	var projectConfigs kargoapi.ProjectConfigList
+	err := s.client.List(
+		ctx,
+		&projectConfigs,
+		client.MatchingFields{
+			indexer.ProjectConfigsByWebhookReceiverPathsField: r.URL.Path,
+		},
+	)
+	if err != nil {
+		logger.Error(err, "failed to list project configs")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if len(projectConfigs.Items) == 0 {
+		logger.Info("no project configs found")
+		http.Error(w, "no project configs found for the request", http.StatusNotFound)
+		return
+	}
+
+	// Projects are allowed to have, at most, a single ProjectConfig
+	pc := projectConfigs.Items[0]
+
+	wrc, err := s.getWebhookReceiverConfig(r.URL.Path, pc)
+	if err != nil {
+		logger.Error(err, "failed to find webhook receiver config")
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+
+	logger.Debug("webhook receiver config found", "webhook-receiver-config", wrc)
+	switch {
+	case wrc.GitHub != nil:
+		var secret corev1.Secret
```

Comment:
I would move all this Secret retrieval logic into the github handler. This way this switch doesn't get out of control when we start directing traffic to many different platform-specific handlers.